lib/orderbook/OrderBook.ts

```diff
@@ -479,51 +479,94 @@ class OrderBook extends EventEmitter {
+      if (remainingQuantityToRemove === 0) {
+        // we can stop listening for swaps once all holds are cleared
+        this.swaps!.removeListener('swap.failed', failedHandler);
+        this.swaps!.removeListener('swap.paid', paidHandler);
+      }
+    };
 
-      const failedHandler = (deal: SwapDeal) => {
-        if (deal.orderId === orderIdentifier.id) {
-          // remove the portion that failed now that it's not on hold
-          this.removeOwnOrder(orderIdentifier.id, orderIdentifier.pairId, deal.quantity!);
-          cleanup(deal.quantity!);
-        }
-      };
+    if (quantityToRemove) {
+      if (quantityToRemove > order.quantity) {
+        // quantity to be removed can't be higher than order's quantity.
+        throw errors.QUANTITY_DOES_NOT_MATCH(quantityToRemove, order.quantity);
+      } else if (quantityToRemove === order.quantity) {
+        this.proceedRemovingOwnOrder(orderIdentifier, order, localId);
+        return;
+      }
+
+      const removableQuantity = order.quantity - order.hold;
+      if (quantityToRemove <= removableQuantity) {
+        this.removeOwnOrder(orderIdentifier.id, orderIdentifier.pairId, quantityToRemove);
+      } else {
+        this.removeOwnOrder(orderIdentifier.id, orderIdentifier.pairId, removableQuantity);
+        remainingQuantityToRemove = quantityToRemove - removableQuantity;
+
+        failedHandler = (deal: SwapDeal) => {
+          if (deal.orderId === orderIdentifier.id) {
+            // remove the portion that failed now that it's not on hold
+            if (deal.quantity! > remainingQuantityToRemove) {
+              this.removeOwnOrder(orderIdentifier.id, orderIdentifier.pairId, remainingQuantityToRemove);
+              cleanup(remainingQuantityToRemove);
+            } else {
+              this.removeOwnOrder(orderIdentifier.id, orderIdentifier.pairId, deal.quantity!);
+              cleanup(deal.quantity!);
+            }
+          }
+        };
 
-      const paidHandler = (result: SwapSuccess) => {
-        if (result.orderId === orderIdentifier.id) {
-          cleanup(result.quantity);
+        this.swaps!.on('swap.failed', failedHandler);
+        this.swaps!.on('swap.paid', paidHandler);
+      }
+    } else {
+      if (order.hold) {
+        remainingQuantityToRemove = order.hold;
+        // we can't remove the entire order as some of it is on hold, start by removing any available portion
+        this.logger.debug(`can't remove local order ${localId} yet because it has a hold of ${order.hold}`);
+        const availableQuantity = order.quantity - order.hold;
+        if (availableQuantity) {
+          this.removeOwnOrder(orderIdentifier.id, orderIdentifier.pairId, availableQuantity);
         }
-      };
 
-      this.swaps!.on('swap.failed', failedHandler);
-      this.swaps!.on('swap.paid', paidHandler);
+        failedHandler = (deal: SwapDeal) => {
```

Comment:
Per my comment above, I'd suggest dropping this additional `failedHandler` definition and only use the one defined above that checks against `remainingQuantityToRemove`, and move it back near the top of the method as a `const`.